src/Microsoft.AspNet.Authentication.Facebook/Notifications/FacebookAuthenticationNotifications.cs

```diff
@@ -23,16 +24,33 @@ public FacebookAuthenticationNotifications()
         /// <summary>
         /// Gets or sets the function that is invoked when the Authenticated method is invoked.
         /// </summary>
-        public Func<FacebookAuthenticatedContext, Task> OnAuthenticated { get; set; }
+        public new Func<FacebookAuthenticatedContext, Task> OnAuthenticated { get; set; }
```

Comment:
Actually, it's not that ugly (specially if we remove the TNotifications generic parameter), but it's not semantically perfect:

If we replace `OAuthAuthenticationNotifications` by `OAuthAuthenticationNotifications<TUser>` and `Func<OAuthAuthenticatedContext, Task> OnAuthenticated` by `Func<OAuthAuthenticatedContext<TUser>, Task> OnAuthenticated`, it implies that we necessarily have a "User" property on `OAuthAuthenticatedContext`, which is not the case if the OAuth2 provider doesn't offer a user-info endpoint. Not to mention that when you don't override the default OAuth2 handler and use the delegate approach, you simply can't fill a user object before the `Authenticated` notification is actually triggered and handled in Startup.cs.

With generics, the "best" approach might be to take a new `TAuthenticatedContext` (hum) generic parameter, change `OnAuthenticated`'s signature to `Func<TAuthenticatedContext, Task> OnAuthenticated`, and expose provider-specific stuff in `*AuthenticatedContext` subclasses, as we do today.

@HaoK @Tratcher your thoughts?